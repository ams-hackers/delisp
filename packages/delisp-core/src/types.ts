//
// Types
//

interface TVoid {
  type: "void";
}

interface TBoolean {
  type: "boolean";
}
interface TNumber {
  type: "number";
}

interface TString {
  type: "string";
}

export interface TApplication {
  type: "application";
  op: string;
  args: Monotype[];
}

export interface TVar {
  type: "type-variable";
  name: string;
}

interface EmptyRow {
  type: "empty-row";
}

interface RowExtension {
  type: "row-extension";
  label: string;
  labelType: Monotype;
  // TODO: Implement kind system!
  extends: Monotype;
}

type Row = EmptyRow | RowExtension;

export type Monotype =
  | TBoolean
  | TNumber
  | TString
  | TApplication
  | TVar
  | Row
  | TVoid;

export interface Type {
  type: "type";
  tvars: string[];
  mono: Monotype;
}

//
// Constructor helpers
//

export const tVoid: TVoid = {
  type: "void"
};

export const tBoolean: TBoolean = {
  type: "boolean"
};

export const tNumber: TNumber = {
  type: "number"
};

export const tString: TString = {
  type: "string"
};

export function tVar(name: string): TVar {
  return {
    type: "type-variable",
    name
  };
}

export function tApp(op: string, ...args: Monotype[]): Monotype {
  return {
    type: "application",
    op,
    args
  };
}

export function tVector(t: Monotype): Monotype {
  return tApp("vector", t);
}

export function tFn(args: Monotype[], out: Monotype): Monotype {
  return tApp("->", ...args, out);
}

const emptyRow: EmptyRow = { type: "empty-row" };

export const tRowExtension = (
  row: Monotype,
  label: string,
  labelType: Monotype
): Row => ({
  type: "row-extension",
  label,
  labelType,
  extends: row
});

export function tRecord(fields: { [key: string]: Monotype }): Monotype {
  return tApp(
    "record",
    Object.keys(fields).reduce(
      (row: Row, label: string): Row =>
        tRowExtension(row, label, fields[label]),
      emptyRow
    )
  );
}
