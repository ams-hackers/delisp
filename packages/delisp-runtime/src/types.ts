export interface Primitives {
  [name: string]: { type: string; value: unknown };
}
