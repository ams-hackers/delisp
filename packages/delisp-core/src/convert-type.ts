import { parseRecord } from "./convert-utils";
import { printHighlightedExpr } from "./error-report";
import {
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";
import { generateUniqueTVar } from "./type-generate";
import {
  emptyRow,
  Monotype,
  tApp,
  tBoolean,
  tNumber,
  tRecord,
  tString,
  tUserDefined,
  tVar,
  tVector,
  tVoid
} from "./types";

/** Capitalize a string like "foo" to "Foo". */
function capitalize(str: string) {
  if (str === "") {
    return "";
  } else {
    return str[0].toUpperCase() + str.slice(1);
  }
}

/** Return true if a symbol is a valid name for a user defined type, false otherwise. */
export function userDefinedType(expr: ASExprSymbol): boolean {
  return expr.name[0] === expr.name[0].toUpperCase();
}

/** Check if a symbol is a valid user defined type name or throw a user-friendly error otherwise. */
export function checkUserDefinedTypeName(expr: ASExprSymbol): void {
  if (!userDefinedType(expr)) {
    throw new Error(
      printHighlightedExpr(
        `'${
          expr.name
        }' is not a valid name for a type. Type names should start with a capital letter. Try '${capitalize(
          expr.name
        )}'?`,
        expr.location
      )
    );
  }
}

function convertSymbol(expr: ASExprSymbol): Monotype {
  switch (expr.name) {
    case "boolean":
      return tBoolean;
    case "number":
      return tNumber;
    case "string":
      return tString;
    case "void":
      return tVoid;
    case "_":
      return generateUniqueTVar(false, "__t");
    default:
      return userDefinedType(expr) ? tUserDefined(expr.name) : tVar(expr.name);
  }
}

function convertList(expr: ASExprList): Monotype {
  const [op, ...args] = expr.elements;

  if (op.type !== "symbol") {
    throw new Error(
      printHighlightedExpr("Expected symbol as operator", expr.location)
    );
  }

  switch (op.name) {
    case "->":
      if (args.length < 1) {
        throw new Error(
          printHighlightedExpr(
            "Expected at least 1 argument",
            op.location,
            true
          )
        );
      }
      break;
    default:
      throw new Error(
        printHighlightedExpr("Unknown type constructor", op.location)
      );
  }

  return tApp(op.name, ...args.map(convert));
}

function convertVector(expr: ASExprVector): Monotype {
  if (expr.elements.length !== 1) {
    throw new Error(
      printHighlightedExpr("Expected exactly 1 argument", expr.location)
    );
  }
  return tVector(convert(expr.elements[0]));
}

function convertMap(expr: ASExprMap): Monotype {
  const { fields, tail } = parseRecord(expr);

  return tRecord(
    fields.map(({ label, value }) => ({
      label: label.name,
      type: convert(value)
    })),
    tail ? convert(tail) : emptyRow
  );
}

/* Try to convert a S-Expression into a type. */
export function convert(expr: ASExpr): Monotype {
  switch (expr.type) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
    default:
      throw new Error(printHighlightedExpr("Not a valid type", expr.location));
  }
}
