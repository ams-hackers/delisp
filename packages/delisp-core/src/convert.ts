import { assertNever } from "./invariant";
import { Location } from "./input";
import { printHighlightedExpr } from "./error-report";
import {
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";
import {
  isExpression,
  Identifier,
  Declaration,
  Expression,
  SLetBindingF,
  LambdaList,
  Syntax
} from "./syntax";
import { foldExpr, exprFChildren } from "./syntax-utils";
import { flatten, last } from "./utils";

import {
  checkUserDefinedTypeName,
  convert as convertType
} from "./convert-type";
import { ConvertError, parseRecord } from "./convert-utils";
import { listTypeVariables } from "./type-utils";
import { TypeWithWildcards } from "./type-wildcards";

export interface WithErrors {
  errors: string[];
}

type ExpressionWithErrors = Expression<WithErrors>;
type DeclarationWithErrors = Declaration<WithErrors, WithErrors>;
type SyntaxWithErrors = Syntax<WithErrors, WithErrors>;

function result(
  node: ExpressionWithErrors["node"],
  location: Location,
  errors: string[]
): ExpressionWithErrors {
  return { node, info: { errors }, location };
}

function success(
  node: ExpressionWithErrors["node"],
  location: Location
): ExpressionWithErrors {
  return result(node, location, []);
}

const conversions: Map<
  string,
  (expr: ASExprList) => ExpressionWithErrors
> = new Map();
const toplevelConversions: Map<
  string,
  (expr: ASExprList) => DeclarationWithErrors
> = new Map();

function defineConversion(
  name: string,
  fn: (expr: ASExprList) => ExpressionWithErrors
) {
  conversions.set(name, fn);
}

function defineToplevel(
  name: string,
  fn: (expr: ASExprList) => DeclarationWithErrors
) {
  toplevelConversions.set(name, fn);
}

function parseBody(anchor: ASExpr, exprs: ASExpr[]): ExpressionWithErrors[] {
  if (exprs.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(`body can't be empty`, anchor.location, true)
    );
  }
  return exprs.map(convertExpr);
}

//
// The format of lambda lists are (a b c ... &rest z)
//

function parseLambdaList(ll: ASExpr): LambdaList {
  if (ll.tag !== "list") {
    throw new ConvertError(
      printHighlightedExpr("Expected a list of arguments", ll.location)
    );
  }

  ll.elements.forEach(arg => {
    if (arg.tag !== "symbol") {
      throw new ConvertError(
        printHighlightedExpr(
          "A list of arguments should be made of symbols",
          arg.location
        )
      );
    }
  });

  const symbols = ll.elements as ASExprSymbol[];

  // Check for duplicated arguments
  symbols.forEach((arg, i) => {
    const duplicated = symbols.slice(i + 1).find(a => a.name === arg.name);
    if (duplicated) {
      throw new ConvertError(
        printHighlightedExpr(
          "There is another argument with the same name",
          duplicated.location
        )
      );
    }
  });

  return {
    positionalArgs: symbols.map(parseIdentifier),
    location: ll.location
  };
}

defineConversion("lambda", expr => {
  const [lambda, ...args] = expr.elements;
  const lastExpr = last([lambda, ...args]) as ASExpr; // we kj

  if (args.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(
        `'lambda' is missing the argument list`,
        lastExpr.location,
        true
      )
    );
  }

  const body = parseBody(lastExpr, args.slice(1));

  return success(
    {
      tag: "function",
      lambdaList: parseLambdaList(args[0]),
      body
    },
    expr.location
  );
});

function missingFrom(expr: ASExpr): ExpressionWithErrors {
  const location = {
    ...expr.location,
    start: expr.location.end,
    end: expr.location.end
  };
  return success({ tag: "unknown" }, location);
}

function exactArguments(
  args: ASExpr[],
  options: {
    required: number;
    fewArguments: string;
    manyArguments: string;
  }
): string[] {
  if (args.length < options.required) {
    return [
      printHighlightedExpr(options.fewArguments, last(args)!.location, true)
    ];
  }
  if (args.length > options.required) {
    return [
      printHighlightedExpr(
        options.fewArguments,
        args[options.required].location,
        true
      )
    ];
  }
  return [];
}

defineConversion("if", expr => {
  const [, conditionForm, consequentForm, alternativeForm] = expr.elements;

  const condition = conditionForm
    ? convertExpr(conditionForm)
    : missingFrom(expr);
  const consequent = consequentForm
    ? convertExpr(consequentForm)
    : missingFrom(expr);
  const alternative = alternativeForm
    ? convertExpr(alternativeForm)
    : missingFrom(expr);

  const receivedArgs = expr.elements.length - 1;

  const errors = exactArguments(expr.elements, {
    required: 4,
    fewArguments: `'if' needs exactly 3 arguments, got ${receivedArgs}`,
    manyArguments: `'if' needs exactly 3 arguments, got ${receivedArgs}`
  });

  return result(
    {
      tag: "conditional",
      condition,
      consequent,
      alternative
    },
    expr.location,
    errors
  );
});

function parseLetBindings(
  bindings: ASExpr
): Array<SLetBindingF<ExpressionWithErrors>> {
  if (bindings.tag !== "map") {
    throw new ConvertError(
      printHighlightedExpr(`'let' bindings should be a map`, bindings.location)
    );
  }
  return bindings.fields.map(field => ({
    variable: parseIdentifier(field.label),
    value: convertExpr(field.value)
  }));
}

defineConversion("let", expr => {
  const [_let, ...args] = expr.elements;
  const lastExpr = last([_let, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(
        `'let' is missing the bindings`,
        lastExpr.location,
        true
      )
    );
  }

  const [rawBindings, ...rawBody] = args;

  return success(
    {
      tag: "let-bindings",
      bindings: parseLetBindings(rawBindings),
      body: parseBody(lastExpr, rawBody)
    },
    expr.location
  );
});

defineConversion("the", expr => {
  const [_the, ...args] = expr.elements;
  const lastExpr = last([_the, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(
        `'the' is missing the type and value`,
        lastExpr.location,
        true
      )
    );
  }
  if (args.length === 1) {
    throw new ConvertError(
      printHighlightedExpr(
        `'the' is missing the expression`,
        lastExpr.location,
        true
      )
    );
  }

  if (args.length > 2) {
    throw new ConvertError(
      printHighlightedExpr(
        `Too many arguments. 'the' should take two arguments.`,
        args[2].location,
        true
      )
    );
  }

  const [t, value] = args;

  return success(
    {
      tag: "type-annotation",
      typeWithWildcards: new TypeWithWildcards(convertType(t)),
      value: convertExpr(value)
    },
    expr.location
  );
});

defineConversion("do", expr => {
  const [_do, ...args] = expr.elements;
  const lastExpr = last([_do, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(`empty body`, lastExpr.location, true)
    );
  }
  const middleForms = args.slice(0, -1);
  const lastForm = last(args)!;
  return success(
    {
      tag: "do-block",
      body: middleForms.map(convertExpr),
      returning: convertExpr(lastForm)
    },
    expr.location
  );
});

defineToplevel("define", expr => {
  const [define, ...args] = expr.elements;

  if (args.length !== 2) {
    const lastExpr = last([define, ...args]) as ASExpr;
    throw new ConvertError(
      printHighlightedExpr(
        `'define' needs exactly 2 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [variable, value] = args;

  if (variable.tag !== "symbol") {
    throw new ConvertError(
      printHighlightedExpr("'define' expected a symbol", variable.location)
    );
  }

  return {
    node: {
      tag: "definition",
      variable: parseIdentifier(variable),
      value: convertExpr(value)
    },
    location: expr.location,
    info: {
      errors: []
    }
  };
});

defineToplevel("export", expr => {
  const [exp, ...args] = expr.elements;

  if (args.length !== 1) {
    const lastExpr = last([exp, ...args]) as ASExpr;
    throw new ConvertError(
      printHighlightedExpr(
        `'export' needs exactly 1 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [variable] = args;

  if (variable.tag !== "symbol") {
    throw new ConvertError(
      printHighlightedExpr("'export' expected a symbol", variable.location)
    );
  }

  return {
    node: {
      tag: "export",
      value: parseIdentifier(variable)
    },
    location: expr.location,
    info: {
      errors: []
    }
  };
});

defineToplevel("type", expr => {
  const [typeOp, ...args] = expr.elements;

  if (args.length !== 2) {
    const lastExpr = last([typeOp, ...args]) as ASExpr;
    throw new ConvertError(
      printHighlightedExpr(
        `'type' needs exactly 2 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [name, definition] = args;

  if (name.tag !== "symbol") {
    throw new ConvertError(
      printHighlightedExpr("'type' expected a symbol as a name", name.location)
    );
  }

  checkUserDefinedTypeName(name);

  const definitionType = convertType(definition);
  if (listTypeVariables(definitionType).length > 0) {
    throw new ConvertError(
      printHighlightedExpr("Type variable out of scope", definition.location)
    );
  }

  return {
    node: {
      tag: "type-alias",
      alias: parseIdentifier(name),
      definition: definitionType
    },
    location: expr.location,
    info: {
      errors: []
    }
  };
});

function convertList(list: ASExprList): ExpressionWithErrors {
  if (list.elements.length === 0) {
    throw new ConvertError(
      printHighlightedExpr("Empty list is not a function call", list.location)
    );
  }

  const [first] = list.elements;

  const convertSpecialForm =
    first.tag === "symbol" ? conversions.get(first.name) : undefined;

  if (convertSpecialForm) {
    return convertSpecialForm(list);
  } else {
    const [fn, ...args] = list.elements;
    return success(
      {
        tag: "function-call",
        fn: convertExpr(fn),
        args: args.map(convertExpr)
      },
      list.location
    );
  }
}

function convertVector(list: ASExprVector): ExpressionWithErrors {
  return success(
    {
      tag: "vector",
      values: list.elements.map(convertExpr)
    },
    list.location
  );
}

function convertMap(map: ASExprMap): ExpressionWithErrors {
  const { fields, tail } = parseRecord(map);
  return success(
    {
      tag: "record",
      fields: fields.map(f => ({
        label: parseIdentifier(f.label),
        value: convertExpr(f.value)
      })),
      extends: tail && convertExpr(tail)
    },
    map.location
  );
}

function parseIdentifier(expr: ASExprSymbol): Identifier {
  return {
    tag: "identifier",
    name: expr.name,
    location: expr.location
  };
}

function convertSymbol(expr: ASExprSymbol): ExpressionWithErrors {
  const id = parseIdentifier(expr);
  return success(
    {
      tag: "variable-reference",
      name: id.name
    },
    id.location
  );
}

function convertExprOrError(expr: ASExpr): ExpressionWithErrors {
  switch (expr.tag) {
    case "number":
      return success({ tag: "number", value: expr.value }, expr.location);
    case "string":
      return success({ tag: "string", value: expr.value }, expr.location);
    case "symbol":
      return convertSymbol(expr);
    case "list":
      return convertList(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
  }
}

function convertExpr(expr: ASExpr): ExpressionWithErrors {
  try {
    return convertExprOrError(expr);
  } catch (err) {
    if (err instanceof ConvertError) {
      return result({ tag: "unknown" }, expr.location, [err.message]);
    } else {
      throw err;
    }
  }
}

function convertOrError(form: ASExpr): SyntaxWithErrors {
  if (form.tag === "list") {
    if (form.elements.length === 0) {
      throw new ConvertError(
        printHighlightedExpr("Empty list is not a function call", form.location)
      );
    }

    const [first] = form.elements;
    const convertDeclaration =
      first.tag === "symbol" ? toplevelConversions.get(first.name) : undefined;

    if (convertDeclaration) {
      return convertDeclaration(form);
    }
  }

  return convertExpr(form);
}

export function convert(form: ASExpr): SyntaxWithErrors {
  try {
    return convertOrError(form);
  } catch (err) {
    if (err instanceof ConvertError) {
      return result({ tag: "unknown" }, form.location, [err.message]);
    } else {
      throw err;
    }
  }
}

//
// Extracing errors
//

function collectConvertExprErrors(expr: ExpressionWithErrors): string[] {
  return foldExpr(expr, e => [...e.info.errors, ...flatten(exprFChildren(e))]);
}

export function collectConvertErrors(syntax: SyntaxWithErrors): string[] {
  if (isExpression(syntax)) {
    return collectConvertExprErrors(syntax);
  } else {
    switch (syntax.node.tag) {
      case "export":
      case "type-alias":
        return syntax.info.errors;
      case "definition":
        return [
          ...syntax.info.errors,
          ...collectConvertExprErrors(syntax.node.value)
        ];
      default:
        return assertNever(syntax.node);
    }
  }
}
