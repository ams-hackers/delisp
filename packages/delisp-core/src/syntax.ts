//
// This module is intended to be imported like `import * as S from "..."`
//

import { Location } from "./input";
import { TypeWithWildcards } from "./type-wildcards";
import { Type } from "./types";

//
// Expressions
//

interface SNumberF {
  tag: "number";
  value: number;
}

interface SStringF {
  tag: "string";
  value: string;
}

interface SBooleanF {
  tag: "boolean";
  value: boolean;
}

interface SNoneF {
  tag: "none";
}

export interface Identifier {
  tag: "identifier";
  name: SVar;
  location: Location;
}

type SVar = string;
interface SVariableReferenceF {
  tag: "variable-reference";
  name: SVar;
}

interface SConditionalF<E> {
  tag: "conditional";
  condition: E;
  consequent: E;
  alternative: E;
}

interface SFunctionCallF<E> {
  tag: "function-call";
  fn: E;
  // User provided function arguments for this function call. This is
  // internal. Use functions from syntax-functions to get this
  // information.
  userArguments: E[];
}

export interface LambdaList {
  tag: "function-lambda-list";
  // The user-provided arguments for the lambda-list. This is
  // internal. Use functions from syntax-function to get this
  // information.
  userPositionalArguments: Identifier[];
  location: Location;
}

interface SFunctionF<E> {
  tag: "function";
  lambdaList: LambdaList;
  body: E[];
}

interface SVectorConstructorF<E> {
  tag: "vector";
  values: E[];
}

export interface SLetBindingF<E> {
  tag: "let-bindings-binding";
  variable: Identifier;
  value: E;
}

interface SLetF<E> {
  tag: "let-bindings";
  bindings: Array<SLetBindingF<E>>;
  body: E[];
}

interface SRecordF<E> {
  tag: "record";
  fields: Array<{
    label: Identifier;
    value: E;
  }>;
  source?: {
    extending: boolean;
    expression: E;
  };
}

interface SRecordGetF<E> {
  tag: "record-get";
  field: Identifier;
  value: E;
}

interface STypeAnnotationF<E> {
  tag: "type-annotation";
  value: E;
  typeWithWildcards: TypeWithWildcards;
}

interface SDoBlockF<E> {
  tag: "do-block";
  body: E[];
  returning: E;
}

export interface SMatchCaseF<E> {
  label: string;
  variable: Identifier;
  body: E[];
}

interface SMatchF<E> {
  tag: "match";
  value: E;
  cases: Array<SMatchCaseF<E>>;
  defaultCase?: E[];
}

interface SCaseTagF<E> {
  tag: "case";
  label: string;
  value?: E;
}

interface SValuesF<E> {
  tag: "values";
  values: E[];
}

interface SMultipleValueBindF<E> {
  tag: "multiple-value-bind";
  variables: Identifier[];
  form: E;
  body: E[];
}

interface SUnknownF<_E> {
  tag: "unknown";
}

type AnyExpressionF<I = {}, E = Expression<I>> =
  | SNumberF
  | SStringF
  | SBooleanF
  | SNoneF
  | SVariableReferenceF
  | SConditionalF<E>
  | SFunctionCallF<E>
  | SFunctionF<E>
  | SVectorConstructorF<E>
  | SLetF<E>
  | SRecordF<E>
  | SRecordGetF<E>
  | STypeAnnotationF<E>
  | SDoBlockF<E>
  | SMatchF<E>
  | SCaseTagF<E>
  | SValuesF<E>
  | SMultipleValueBindF<E>
  | SUnknownF<E>;

interface Node<I, E> {
  node: E;
  location: Location;
  info: I;
}

export interface ExpressionF<I, E> extends Node<I, AnyExpressionF<I, E>> {}

export interface Expression<I = {}>
  extends Node<I, AnyExpressionF<I, Expression<I>>> {}

export interface SVariableReference<I = {}>
  extends Node<I, SVariableReferenceF> {}

export interface SConditional<I = {}>
  extends Node<I, SConditionalF<Expression<I>>> {}

export interface SFunctionCall<I = {}>
  extends Node<I, SFunctionCallF<Expression<I>>> {}

export interface SFunction<I = {}> extends Node<I, SFunctionF<Expression<I>>> {}

export interface SLet<I = {}> extends Node<I, SLetF<Expression<I>>> {}

export interface SRecord<I = {}> extends Node<I, SRecordF<Expression<I>>> {}
export interface SRecordGet<I = {}>
  extends Node<I, SRecordGetF<Expression<I>>> {}

export interface SVectorConstructor<I = {}>
  extends Node<I, SVectorConstructorF<Expression<I>>> {}

export interface SDoBlock<I = {}> extends Node<I, SDoBlockF<Expression<I>>> {}

export interface SMatch<I = {}> extends Node<I, SMatchF<Expression<I>>> {}

export interface SCaseTag<I = {}> extends Node<I, SCaseTagF<Expression<I>>> {}

export interface SValues<I = {}> extends Node<I, SValuesF<Expression<I>>> {}

export interface SMultipleValueBind<I = {}>
  extends Node<I, SMultipleValueBindF<Expression<I>>> {}

export interface SUnknown<I = {}> extends Node<I, SUnknownF<Expression<I>>> {}

//
// Declarations
//

interface SDefinitionF<E> {
  tag: "definition";
  variable: Identifier;
  value: E;
}
export interface SDefinition<EInfo = {}, SInfo = {}>
  extends Node<SInfo, SDefinitionF<Expression<EInfo>>> {}

interface SImportF<_E> {
  tag: "import";
  variable: Identifier;
  source: string;
}
export interface SImport<I = {}> extends Node<I, SImportF<Expression<I>>> {}

interface SExportF<_E> {
  tag: "export";
  identifiers: Identifier[];
}
export interface SExport<I = {}> extends Node<I, SExportF<Expression<I>>> {}

interface STypeAliasF<_E> {
  tag: "type-alias";
  alias: Identifier;
  definition: Type;
}
export interface STypeAlias<I = {}>
  extends Node<I, STypeAliasF<Expression<I>>> {}

export type Declaration<EInfo = {}, SInfo = {}> =
  | SDefinition<EInfo, SInfo>
  | SExport<SInfo>
  | SImport<SInfo>
  | STypeAlias<SInfo>;

export type Syntax<EInfo = {}, SInfo = {}> =
  | Expression<EInfo>
  | Declaration<EInfo, SInfo>;

export function isDeclaration<I>(syntax: Syntax<I>): syntax is Declaration<I> {
  return (
    syntax.node.tag === "definition" ||
    syntax.node.tag === "import" ||
    syntax.node.tag === "export" ||
    syntax.node.tag === "type-alias"
  );
}

export function isExpression<I>(syntax: Syntax<I>): syntax is Expression<I> {
  return !isDeclaration(syntax);
}

export function isDefinition<I>(syntax: Syntax<I>): syntax is SDefinition<I> {
  return syntax.node.tag === "definition";
}

export function isImport<I>(syntax: Syntax<I>): syntax is SImport<I> {
  return syntax.node.tag === "import";
}

export function isExport<I>(syntax: Syntax<I>): syntax is SExport<I> {
  return syntax.node.tag === "export";
}

export function isTypeAlias<I>(syntax: Syntax<I>): syntax is STypeAlias<I> {
  return syntax.node.tag === "type-alias";
}

export interface Module<EInfo = {}, SInfo = {}> {
  tag: "module";
  body: Array<Syntax<EInfo, SInfo>>;
}
