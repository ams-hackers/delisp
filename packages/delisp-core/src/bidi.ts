import { InvariantViolation, assertNever } from "./invariant";
import { Expression } from "./syntax";
import { Type, tString, tNumber } from "./types";
import { unify } from "./unify";
import { printType } from "./type-printer";
import { last } from "./utils";
import { Location } from "./input";
import { printHighlightedExpr } from "./error-report";

export interface Environment {
  [variable: string]: Type;
}

function equalTypes(t1: Type, t2: Type): boolean {
  // Quick and dirty equality definition for trying out.
  const result = unify(t1, t2, {});
  return (
    result.tag === "unify-success" &&
    Object.keys(result.substitution).length === 0
  );
}

function typeError(message: string, entity: { location: Location }) {
  return new Error(printHighlightedExpr(message, entity.location));
}

export function infer(expr: Expression, env: Environment): Type | null {
  switch (expr.node.tag) {
    case "string":
      return tString;
    case "number":
      return tNumber;
    case "variable-reference": {
      const variableName = expr.node.name;
      const variableType = env[variableName];
      if (!variableType) {
        throw typeError(`Unknown variable ${variableName}`, expr);
      }
      return variableType;
    }
    case "type-annotation": {
      const typ = expr.node.typeWithWildcards.instantiate();
      check(expr.node.value, typ, env);
      return typ;
    }
    case "function-call": {
      const fnType = infer(expr.node.fn, env);
      if (!fnType) {
        throw typeError(
          `Can't called a function with unknown type.`,
          expr.node.fn
        );
      }

      if (
        !(
          fnType.node.tag === "application" &&
          fnType.node.op.node.tag === "constant" &&
          fnType.node.op.node.name === "->"
        )
      ) {
        throw typeError(
          `Looks like you are trying to do a function call on a ${printType(
            fnType
          )}`,
          expr.node.fn
        );
      }

      const argTypes = fnType.node.args.slice(0, -1);
      const outType = last(fnType.node.args);
      if (!outType) {
        throw new InvariantViolation(`Function with no output type?`);
      }

      expr.node.args.forEach((arg, i) => {
        check(arg, argTypes[i], env);
      });

      return outType;
    }

    default:
      return null;
  }
}

export function check(expr: Expression, type: Type, env: Environment) {
  switch (expr.node.tag) {
    case "string":
    case "number":
    case "variable-reference":
    case "type-annotation":
    case "function-call": {
      const inferred = infer(expr, env);
      if (inferred) {
        if (!equalTypes(inferred, type)) {
          throw typeError(
            `Expected type ${printType(type, false)}, but found ${printType(
              inferred
            )}.`,
            expr
          );
        }
      }
      return;
    }
    case "function": {
      if (
        !(
          type.node.tag === "application" &&
          type.node.op.node.tag === "constant" &&
          type.node.op.node.name === "->"
        )
      ) {
        throw typeError(
          `Expected a function, but found ${printType(type)}`,
          expr
        );
      }

      const argTypes = type.node.args.slice(0, -1);
      const outType = last(type.node.args);
      if (!outType) {
        throw new InvariantViolation(`Function with no output type?`);
      }

      const args = expr.node.lambdaList.positionalArgs.map(i => i.name);

      if (argTypes.length !== args.length) {
        throw typeError(
          `Wrong number of arguments. Expected ${argTypes.length} but found ${
            args.length
          }`,
          expr.node.lambdaList
        );
      }

      const newEnv = args.reduce(
        (tmpEnv, a, idx) => ({
          ...tmpEnv,
          [a]: argTypes[idx]
        }),
        env
      );

      check(last(expr.node.body)!, outType, newEnv);
      return;
    }
    case "conditional":
    case "let-bindings":
    case "record":
    case "vector":
      throw typeError(
        `Cannot check this expression for type ${printType(type)}.`,
        expr
      );
    default:
      assertNever(expr.node);
  }
}
