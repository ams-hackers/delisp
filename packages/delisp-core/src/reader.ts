//
// reader.ts --- The reader for Delisp
//

// import { List, Syntax } from "./syntax";

//
// S-Expression parser
//

import {
  Parser,
  regex,
  alternatives,
  many,
  delimited
} from "./parser-combinators";

import { ASExpr } from "./syntax";

const spaces = regex(/^\s*/);

function spaced<A>(x: Parser<A>) {
  return delimited(spaces, x, spaces);
}

const number: Parser<ASExpr> = regex(/[0-9]+/)
  .map(
    (string, location): ASExpr => ({
      type: "number",
      value: parseInt(string, 10),
      location
    })
  )
  .description("number");

const doubleQuote = regex(/"/).description("double quote");

const stringChar = regex(/[^"\\]/).description("non-escaped string character");
const stringEscapedChar = regex(/\\[n\\]/).description(
  "escaped string character"
);
const stringConsitutent = alternatives(stringChar, stringEscapedChar);

const string = doubleQuote
  .then(many(stringConsitutent))
  .skip(doubleQuote)
  .map(
    (chars, location): ASExpr => ({
      type: "string",
      value: chars.join(""),
      location
    })
  );

const symbol: Parser<ASExpr> = regex(/[a-zA-Z+<>!@$%^*/-]+/)
  .map(
    (name, location): ASExpr => ({
      type: "symbol",
      name,
      location
    })
  )
  .description("symbol");

const atom: Parser<ASExpr> = alternatives(number, string, symbol).description(
  "atom"
);

const leftParen = regex(/\(/).description("open parenthesis");
const rightParen = regex(/\)/).description("close parenthesis");

function list(x: Parser<ASExpr>): Parser<ASExpr> {
  return leftParen
    .then(many(x))
    .skip(spaces)
    .skip(rightParen)
    .map(
      (elements, location): ASExpr => ({
        type: "list",
        elements,
        location
      })
    );
}

const sexpr: Parser<ASExpr> = spaced(atom.or(() => list(sexpr)));

//
// Parser a Delisp expression from a string
//
export function readFromString(str: string) {
  return sexpr.parse(str);
}
