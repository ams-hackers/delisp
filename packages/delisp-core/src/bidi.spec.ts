import { Expression } from "./syntax";
import { Type } from "./types";

import { readFromString } from "./reader";
import { convertExpr } from "./convert";
import { convert as convertType } from "./convert-type";

import { infer, check, Environment } from "./bidi";

function expr(str: string): Expression {
  return convertExpr(readFromString(str));
}

function type(str: string): Type {
  return convertType(readFromString(str));
}

function typeError(str: string, env: Environment = {}) {
  try {
    infer(expr(str), env);
  } catch (err) {
    return err.message;
  }
  throw new Error(`Did not fail`);
}

const env = {
  true: type("boolean"),
  false: type("boolean"),
  "+": type("(-> number number number)")
};

describe("Bidirectional typing", () => {
  test("number types are infered", () => {
    expect(infer(expr("23"), {})).toEqual(type("number"));
  });

  test("number types are checked", () => {
    expect(() => check(expr("23"), type("number"), {})).not.toThrow();
  });

  test("function types are checked", () => {
    expect(() =>
      infer(expr(`(the (-> number number) (lambda (x) x))`), {})
    ).not.toThrow();
  });

  test("function call types are inferred", () => {
    expect(() => infer(expr(`(+ 2 3)`), env)).not.toThrow();
  });

  test("conditionals are infered", () => {
    expect(() =>
      check(expr("(if true 1 2)"), type("number"), env)
    ).not.toThrow();
  });

  describe("Error messages", () => {
    test("annotating a string as a number", () => {
      expect(typeError(`(the number "foo")`)).toMatchSnapshot();
    });

    test("wrong number of declared arguments", () => {
      expect(
        typeError(`(the (-> number number number) (lambda (x) x))`)
      ).toMatchSnapshot();
    });

    test("wrong argument", () => {
      expect(typeError(`(+ 2 "foo")`, env)).toMatchSnapshot();
    });
    test("ambiguous conditional", () => {
      expect(typeError(`(if true 1 "foo")`, env)).toMatchSnapshot();
    });
    test("non-boolean conditional", () => {
      expect(typeError(`(if 1 "abc" "foo")`, env)).toMatchSnapshot();
    });
  });
});
