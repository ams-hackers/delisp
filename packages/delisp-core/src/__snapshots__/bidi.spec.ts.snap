// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bidirectional typing Error messages annotating a string as a number 1`] = `
"file:1:13: Expected type number, but found string.
(the number \\"foo\\")
------------^"
`;

exports[`Bidirectional typing Error messages wrong argument 1`] = `
"file:1:6: Expected type number, but found string.
(+ 2 \\"foo\\")
-----^"
`;

exports[`Bidirectional typing Error messages wrong number of declared arguments 1`] = `
"file:1:40: Wrong number of arguments. Expected 2 but found 1
(the (-> number number number) (lambda (x) x))
---------------------------------------^"
`;
